Refresh BPF related rules in RHEL 9 OSPP profile

jan-cerny opened the pull request:
#### Description:
First, we remove rule sysctl_net_core_bpf_jit_harden from the RHEL 9 OSPP profile.

This rule sets the net.core.bpf_jit_harden value to 2. However, contrary to the information about the values on https://www.kernel.org/doc/html/latest/admin-guide/sysctl/net.html#bpf-jit-harden (0 = disable JIT hardening, 1 = enable JIT hardening for unprivileged users only, 2 = enable JIT hardening for all users), it's clarified that bpf_jit_harden=1 disables kallsyms access from bpf programs and all users, and it turns on constants blinding by using random value + XOR for CAP_BPF; so the only thing in which value 1 and 2 differ is the constants blinding for CAP_SYS_ADMIN processes in the initial user namespaces. The extra constants blinding with bpf_jit_harden=2 does not really help with CVE mitigation.

Second, we replace rule sysctl_kernel_unprivileged_bpf_disabled in the RHEL 9 OSPP profile by a new rule sysctl_kernel_unprivileged_bpf_disabled_accept_default. This new rule is introduced in this PR.

This rule sets the kernel.unprivileged_bpf_disabled value to 1.  However, on RHEL 9 the kernel supports new value 2 which per https://www.kernel.org/doc/html/latest/admin-guide/sysctl/kernel.html#unprivileged-bpf-disabled makes it for a privileged admin to re-enable unprivileged BPF. The value 2 is also the RHEL 9 default. So the current sysctl_kernel_unprivileged_bpf_disabled rule unnecessarily modifies the RHEL 9 default.  Therefore, we'd like the sysctl_kernel_unprivileged_bpf_disabled rule to accept any non-zero value (1, 2). As this rule is used in many other profiles and products where the value 2 isn't available, we will introduce a new similar rule.

The biggest part of this PR is changes making the sysctl template more flexible so that the new rule sysctl_kernel_unprivileged_bpf_disabled_accept_defaul can be created. These changes to sysctl template include:
- use rule ID as a root of element identifiers in the OVAL template instead of using the sysctlvar parameter for generating the element identifiers. Element identifiers are OVAL definition ID, OVAL test ID, OVAL object ID, OVAL state ID, OVAL variable ID etc. This change means that there could be 2 different rules configuring the same sysctl and build won't abort in such situation. Moreover, it's a convention in other templates.
- allow possibility to have multiple values (a list of values) instead of a single value as the `sysctlval` argument of the template
- option to choose one of these values in the `sysctlval` for remediations (new parameter `sysctlval_remediate`), because if we have multiple values on allow list we need to know which one set in the remediation
- add parameter `missing_static_pass` which will cause that the generated OVAL check will pass if the system default is not modified by explicitly set configuration. This allows us to avoid unnecessary system configuration changes.

The PR also adds a few test scenarios for the new rule, these test scenarios test the aforementioned new features of the sysctl template.

#### Rationale:

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=2081728

matejak commented on docs/templates/template_reference.md:
```diff
@@ -815,9 +815,26 @@ The selected value can be changed in the profile (consult the actual variable fo
     -   **datatype** - data type of the sysctl value, eg. `int`.
 
-    -   **sysctlval** - value of the sysctl value, eg. `'1'`. If this
+    -   **sysctlval** - value of the sysctl value. This can be either an atomic
```
A general rant - the sysctl template shouldn't have sysctlval and sysctlvar - those two words are very similar, and the "sysctl" part is implied. It should have "name" and "values". But don't change this in the PR - if we ever decide to change argument names, that should be done separately.

matejak commented on docs/templates/template_reference.md:
```diff
@@ -815,9 +815,26 @@ The selected value can be changed in the profile (consult the actual variable fo
+    -   **sysctlval_wrong** - the value that is always wrong. This will be used
+        only in the test scenarios only if **sysctlval** is a list.
+
+    -   **missing_static_pass** - if set to `true` the check will pass if the
```
Our project already has a notion of "missing_parameter_pass" - I would lean towards unifying the name. I don't find "missing_static_pass" significantly more descriptive.

ggbecker replied:
The intent behind this kind of parameter has always been somewhat controversial. Usually STIG profiles require them to be explicitly set. What we did for the SSHD parameters is that values are put under a different config file under the .d drop-in config directory. Unfortunately we cannot tailor this behavior per profile. For now it should be ok as we don't plan to use these rules across many different profiles, but we need to keep in mind this scenario.

matejak replied:
This is troublesome, sometimes customers want us to make checks [minimalistic](https://bugzilla.redhat.com/show_bug.cgi?id=2077531), sometimes they want them to be explicit (STIG). We could design content that would use an XCCDF value to parametrize this behavior, but that would be quite ugly (and out of the scope of this PR).

ggbecker commented on docs/templates/template_reference.md:
```diff
@@ -815,9 +815,26 @@ The selected value can be changed in the profile (consult the actual variable fo
+        value, eg. `'1'`, or a list of values, eg. `['1','2']`. If this
         parameter is not specified, XCCDF Value is used instead.
 
+    -   **sysctlval_remediate** - the value that will be used in remediations.
```
This could be replaced by an XCCDF variable. The variable would be only used in the remediation and would allow users to tailor the value, while here the value is hardcoded and can be only changed during build time.

ggbecker commented on linux_os/guide/system/permissions/restrictions/sysctl_kernel_unprivileged_bpf_disabled_accept_default/rule.yml:
```diff
@@ -0,0 +1,83 @@
+    name: sysctl
+    vars:
+        sysctlvar: kernel.unprivileged_bpf_disabled
+        sysctlval:
+        - '1'
+        - '2'
+        sysctlval_remediate: "2"
+        sysctlval_wrong: "0"
```
depending on how the design is done we could probably do something like:
```suggestion
        sysctlval:
        - check:
          - '1'
          - '2'
        - remediate: "2"
        - wrong_value: "0"
```

or
similar but with the wrong value on a different level

```
name: sysctl
vars: ...
metadata:
    - test:
        - wrong_value_input: "0"
```

matejak replied:
Although this nesting idea makes sense from the logical point of view, I think that as the amount of values is small, the benefit is limited. I would also consider the cost that content authors will have to deal with this nesting/indentation instead of having a simple linear list, so I would lean towards keeping everything on a same level.

If you decide to go with the nested list, I would think about names and their descriptiveness e.g. use `valid` or `allowed` instead of `check`, descriptive name for the wrong value.

ggbecker replied:
I also wanted to reinforce the `metadata` solution by pointing out that we also have the need to parametrize packages that are installed before the tests execute and this can be the foundation of the job. Although it's definitely out of scope of this PR.

jan-cerny commented on docs/templates/template_reference.md:
```diff
@@ -815,9 +815,26 @@ The selected value can be changed in the profile (consult the actual variable fo
+        should be remedied to. When  the **sysctlval** parameter is not a list
+        this parameter is optional.
+
+    -   **sysctlval_wrong** - the value that is always wrong. This will be used
```
@matejak We can also change the OVAL to add additional branch checking the "wrong" value isn't set and then the parameter wouldn't be a tests-only parameter but a full citizen.

matejak replied:
Not needed now, when we have a list of allowed values, so I would lean against this. However, it may prove to be useful when such need arises.

jan-cerny commented:
I have tried to address the people feedback. Specifically, I have replaced the sysctlval_remediate parameter by using an XCCDF value, renamed sysctlval_wrong to wrong_sysctlval_for_testing and renamed the missing_static_pass parameter to missing_parameter_pass.

yuumasato commented on docs/templates/template_reference.md:
```diff
@@ -815,8 +815,24 @@ The selected value can be changed in the profile (consult the actual variable fo
+            All remediations will use an XCCDF value instead.
+
+    -   **wrong_sysctlval_for_testing** - the value that is always wrong. This will be used
+        only in the templated test scenarios only if **sysctlval** is a list.
```
`only` is used twice in the sentence.
```suggestion
        only in the templated test scenarios when **sysctlval** is a list.
```

yuumasato commented on linux_os/guide/system/permissions/restrictions/sysctl_kernel_unprivileged_bpf_disabled_accept_default/rule.yml:
```diff
@@ -0,0 +1,82 @@
+ocil_clause: "the kernel.unprivileged_bpf_disabled is not set to 1 or 2 or is configured to be 0"
+
+fixtext: |-
+    Configure {{{ full_name }}} to prevent privilege escalation thru the kernel by disabling access to the bpf syscall.
```
```suggestion
    Configure {{{ full_name }}} to prevent privilege escalation through the kernel by disabling access to the bpf syscall.
```

yuumasato commented on linux_os/guide/system/permissions/restrictions/sysctl_kernel_unprivileged_bpf_disabled_accept_default/rule.yml:
```diff
@@ -0,0 +1,82 @@
+    ospp: FMT_SMF_EXT.1
+    srg: SRG-OS-000132-GPOS-00067,SRG-OS-000480-GPOS-00227
+    stigid@ol8: OL08-00-040281
+    stigid@rhel8: RHEL-08-040281
```
I think the STIGID from the other rule be removed.

yuumasato commented on docs/templates/template_reference.md:
```diff
@@ -815,8 +815,24 @@ The selected value can be changed in the profile (consult the actual variable fo
+    -   **sysctlval** - value of the sysctl value. This can be either not
+        specified, or an atomic value, eg. `'1'`, or a list of values,
+        eg. `['1','2']`.
+        -   If this parameter is not specified, an XCCDF Value is used instead
```
Maybe we should clarify that the  XCCDF Value used is based on the name of the sysctl option, i.e.: `sysctlvar`.
It was already not very clear before...

yuumasato reviewed (commented):
Looking pretyyy good, IMO.

jan-cerny commented:
I have improved the documentation according to the review feedback, then, I removed the STIG ID and fixed the grammar.

jan-cerny commented:
/retest

vojtapolasek commented on linux_os/guide/system/permissions/restrictions/sysctl_kernel_unprivileged_bpf_disabled_accept_default/rule.yml:
```diff
@@ -0,0 +1,80 @@
+    however, an admin can still change this setting later on, if needed, by
+    writing <code>0</code> or <code>1</code> to this entry.
+
+    {{{ describe_sysctl_option_value(sysctl="kernel.unprivileged_bpf_disabled", value="1") }}}
```
In the above description, you taslk about values of "1" or "2". But here you describe only value "1".

jan-cerny commented:
I have mention both values 1 and 2 in the rule description.

vojtapolasek commented:
I used Automatus to do a test in template mode for sysctl template. Here are the results from using Bash remediations so far, Ansible will be later. I used RHEL9.
ERROR - Rule 'sysctl_kernel_core_pattern' test setup script 'two_sysctls_on_same_file_name.fail.sh' failed with exit code 1
ERROR - Environment failed to prepare, skipping test

ERROR - Rule 'sysctl_kernel_core_pattern' test setup script 'two_sysctls_on_d.fail.sh' failed with exit code 1
ERROR - Environment failed to prepare, skipping test

ERROR - Rule 'sysctl_kernel_core_pattern' test setup script 'symlink_repeated_sysctl_conf.pass.sh' failed with exit code 1
ERROR - Environment failed to prepare, skipping test

ERROR - Rule 'sysctl_kernel_core_pattern' test setup script 'symlink_root_duplicate.fail.sh' failed with exit code 1
ERROR - Environment failed to prepare, skipping test

ERROR - Rule 'sysctl_kernel_core_pattern' test setup script 'symlinks_to_same_file.pass.sh' failed with exit code 1
ERROR - Environment failed to prepare, skipping test

ERROR - Rule 'sysctl_kernel_core_pattern' test setup script 'two_sysctls_on_same_file.pass.sh' failed with exit code 1
ERROR - Environment failed to prepare, skipping test

ERROR - Rule 'sysctl_kernel_core_pattern' test setup script 'symlink_same_option.fail.sh' failed with exit code 1
ERROR - Environment failed to prepare, skipping test

ERROR - Rule 'sysctl_kernel_core_pattern' test setup script 'symlink_root_incompliant.fail.sh' failed with exit code 1
ERROR - Environment failed to prepare, skipping test

ERROR - Rule 'sysctl_kernel_core_pattern' test setup script 'symlink_different_option.pass.sh' failed with exit code 1
ERROR - Environment failed to prepare, skipping test

ERROR - Script wrong_runtime.fail.sh using profile (all) found issue:
ERROR - Rule evaluation resulted in pass, instead of expected fail during initial stage 
ERROR - The initial scan failed for rule 'xccdf_org.ssgproject.content_rule_sysctl_fs_protected_symlinks'.

ERROR - Script wrong_runtime.fail.sh using profile (all) found issue:
ERROR - Rule evaluation resulted in pass, instead of expected fail during initial stage 
ERROR - The initial scan failed for rule 'xccdf_org.ssgproject.content_rule_sysctl_fs_protected_hardlinks'.

ERROR - Rule 'sysctl_net_ipv4_ip_local_port_range' test setup script 'two_sysctls_on_same_file_name.fail.sh' failed with exit code 255
ERROR - Environment failed to prepare, skipping test

ERROR - Rule 'sysctl_net_ipv4_ip_local_port_range' test setup script 'two_sysctls_on_d.fail.sh' failed with exit code 255
ERROR - Environment failed to prepare, skipping test

ERROR - Rule 'sysctl_net_ipv4_ip_local_port_range' test setup script 'symlink_repeated_sysctl_conf.pass.sh' failed with exit code 255
ERROR - Environment failed to prepare, skipping test

ERROR - Rule 'sysctl_net_ipv4_ip_local_port_range' test setup script 'symlink_root_duplicate.fail.sh' failed with exit code 255
ERROR - Environment failed to prepare, skipping test

ERROR - Rule 'sysctl_net_ipv4_ip_local_port_range' test setup script 'symlinks_to_same_file.pass.sh' failed with exit code 255
ERROR - Environment failed to prepare, skipping test

ERROR - Rule 'sysctl_net_ipv4_ip_local_port_range' test setup script 'two_sysctls_on_same_file.pass.sh' failed with exit code 255
ERROR - Environment failed to prepare, skipping test

ERROR - Rule 'sysctl_net_ipv4_ip_local_port_range' test setup script 'symlink_same_option.fail.sh' failed with exit code 255
ERROR - Environment failed to prepare, skipping test

ERROR - Rule 'sysctl_net_ipv4_ip_local_port_range' test setup script 'symlink_root_incompliant.fail.sh' failed with exit code 255
ERROR - Environment failed to prepare, skipping test

ERROR - Rule 'sysctl_net_ipv4_ip_local_port_range' test setup script 'symlink_different_option.pass.sh' failed with exit code 255
ERROR - Environment failed to prepare, skipping test

None of rules have rule-specific tests.

vojtapolasek commented:
The same holds when remediated with Ansible.

jan-cerny commented:
@vojtapolasek The same problem is reproducible also with the current upstream master branch (as of HEAD ebae31973c00f07e53bbafe71ad9699386885a71) so it probably isn't caused by this PR. I will look into this.

jan-cerny commented:
@vojtapolasek I have tried to fix this in https://github.com/ComplianceAsCode/content/pull/9269.

vojtapolasek commented:
Thanks for the fix of test templates. Can you please rebase on master to see this fix also in test results of this PR?

jan-cerny commented:
I have rebased

vojtapolasek reviewed (approved):
I think it looks good. There is still one failing test when testing sysctl template, it occurs with rule sysctl_fs_protected_symlinks.
But it happens in this PR as well as in master branch. CI failures are caused by the fact that tests are running in container.
On my local machine all sysctl templated rules are working as expected except the one mentioned above.

yuumasato commented on linux_os/guide/system/permissions/restrictions/sysctl_kernel_unprivileged_bpf_disabled_value.var:
```diff
@@ -0,0 +1,18 @@
+
+options:
+    default: 2
+    0: "0"
```
@jan-cerny Do you think it makes sense to offer a selector for value of `0` when the check will never accept it?
